### PR TITLE
dump/: integrate dumpling

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
https://github.com/pingcap/dm/issues/526

### What is changed and how it works?
Dump sql using dumpling. It's not easy to use for users because they need to download a mydumper binary file.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression

Related changes

 - Need to update the documentation
 - Need to update the `dm/dm-ansible`
 - Need to be included in the release note

### CODE REVIEW

**Comment by lichunzhu** [comment]
@WangXiangUSTC @csuzhangxc PTAL

File: tests/all_mode/data/db1.increment.sql [hunk 1]
```diff
@@ -25,7 +25,7 @@ update t1 set name = 'gentestxxxxxx' where gen_id = 124;
 delete from t1 where gen_id > 124;
 
 -- test alter database
-alter database all_mode CHARACTER SET = utf8mb4;
+-- alter database all_mode CHARACTER SET = utf8mb4;
```
Comment (csuzhangxc) [inline tests/all_mode/data/db1.increment.sql]: Is it still not support this?

File: tests/print_status/conf/dm-task.yaml [hunk 2]
```diff
@@ -30,7 +30,8 @@ mydumpers:
     threads: 4
     chunk-filesize: 0
     skip-tz-utc: true
-    extra-args: "--statement-size=5000"
+    statement-size: 4000
```
Comment (csuzhangxc) [inline tests/print_status/conf/dm-task.yaml]: Why change `5000` to `4000` here?

File: tests/sequence_sharding/run.sh [hunk 3]
```diff
@@ -33,7 +33,7 @@ function run() {
     run_sql_file $cur/data/db1.increment.sql $MYSQL_HOST1 $MYSQL_PORT1 $MYSQL_PASSWORD1
     run_sql_file $cur/data/db2.increment.sql $MYSQL_HOST2 $MYSQL_PORT2 $MYSQL_PASSWORD2
 
-    sleep 3
+    sleep 10
```
Comment (csuzhangxc) [inline tests/sequence_sharding/run.sh]: why do we need to change from `3` to `10`?

File: dm/worker/metrics.go [hunk 4]
```diff
@@ -62,6 +63,7 @@ func RegistryMetrics() {
 
 	relay.RegisterMetrics(registry)
 	mydumper.RegisterMetrics(registry)
+	dumpling.RegisterMetrics(registry)
```
Comment (csuzhangxc) [inline dm/worker/metrics.go]: Do we need to remove `mydumper.RegisterMetrics(registry)` now?

File: dumpling/dumpling.go [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+	})
+
+	// NOTE: remove output dir before start dumping
+	// every time re-dump, loader should re-prepare
+	err := os.RemoveAll(m.cfg.Dir)
+	if err != nil {
+		m.logger.Error("fail to remove output directory", zap.String("directory", m.cfg.Dir), log.ShortError(err))
+	}
+
+	// Cmd cannot be reused, so we create a new cmd when begin processing
```
Comment (csuzhangxc) [inline dumpling/dumpling.go]: We can remove this log now.

File: dumpling/dumpling.go [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+		ret = append(ret, "--skip-tz-utc")
+	}
+
+	extraArgs := strings.Fields(cfg.ExtraArgs)
+	if len(extraArgs) > 0 {
+		ret = append(ret, ParseArgLikeBash(extraArgs)...)
+	}
+
+	// TODO: support String for export.Config
+	// m.logger.Info("create dumpling", zap.Stringer("config", dumpConfig))
```
Comment (csuzhangxc) [inline dumpling/dumpling.go]: uncomment this line or just remove it.

File: dumpling/dumpling.go [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+func (m *Dumpling) IsFreshTask(ctx context.Context) (bool, error) {
+	return true, nil
+}
+
+// constructArgs constructs arguments for exec.Command
+func (m *Dumpling) constructArgs() (*export.Config, error) {
+	cfg := m.cfg
+	db := cfg.From
+
+	dumpConfig := export.DefaultConfig()
```
Comment (csuzhangxc) [inline dumpling/dumpling.go]: 1. Can we disable `StatusAddr` now? because we already have it in DM, and if we enable it, may we need more ports for the process.
2. Do we need to set `ServerInfo` based on `flavor`?

File: tests/all_mode/data/db1.increment.sql (hunk 1 above)
Comment (lichunzhu) [reply tests/all_mode/data/db1.increment.sql]: This can't be passed because the default CHARACTER is set to latin1. If we revise this line to
```sql
-- alter database all_mode CHARACTER SET = latin1;
```
I think we can pass it.

File: tests/print_status/conf/dm-task.yaml (hunk 2 above)
Comment (lichunzhu) [reply tests/print_status/conf/dm-task.yaml]: Because the difference of the meaning of statement-size between mydumper and dumpling. dumpling only counts the data's size, not counting the joiners. So I shorten this to split it to more files.

File: dumpling/dumpling.go (hunk 7 above)
Comment (lichunzhu) [reply dumpling/dumpling.go]: 1. I will block it.
2. dumpling will adjust `ServerInfo` during the `Dump` function.

**Comment by lichunzhu** [comment]
https://github.com/pingcap/dumpling/pull/51
dumpling has proposed a PR to combine logger with DM. This PR should include this feature and be merged after this PR is merged.

File: dumpling/dumpling.go [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+		m.logger.Info("dump unit runs forever", zap.String("failpoint", "dumpUnitProcessForever"))
+		<-ctx.Done()
+		failpoint.Return()
+	})
+
+	// NOTE: remove output dir before start dumping
+	// every time re-dump, loader should re-prepare
+	err := os.RemoveAll(m.cfg.Dir)
+	if err != nil {
+		m.logger.Error("fail to remove output directory", zap.String("directory", m.cfg.Dir), log.ShortError(err))
```
Comment (WangXiangUSTC) [inline dumpling/dumpling.go]: do we need to append err to `errs` and then return? If the old dir is not removed, may load wrong data to upstream

File: dumpling/dumpling.go [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+
+	err = export.Dump(m.dumpConfig)
+
+	if err != nil {
+		dumplingExitWithErrorCounter.WithLabelValues(m.cfg.Name).Inc()
+		errs = append(errs, unit.NewProcessError(err))
+	} else {
+		select {
+		case <-ctx.Done():
+			isCanceled = true
```
Comment (WangXiangUSTC) [inline dumpling/dumpling.go]: what is the `isCanceled` used for?

File: dumpling/dumpling.go [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+		dumpConfig.StatementSize = cfg.StatementSize
+	}
+	if cfg.Rows > 0 {
+		dumpConfig.Rows = cfg.Rows
+	}
+	if len(cfg.Where) > 0 {
+		dumpConfig.Where = cfg.Where
+	}
+
+	var ret []string
```
Comment (WangXiangUSTC) [inline dumpling/dumpling.go]: `ret` is used to save some unsupported arguments? how about adding a comment

File: dumpling/util.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
+// Copyright 2019 PingCAP, Inc.
```
Comment (WangXiangUSTC) [inline dumpling/util.go]:
```suggestion
// Copyright 2020 PingCAP, Inc.
```

File: dumpling/dumpling.go [hunk 12]
```diff
@@ -0,0 +1,228 @@
+// Copyright 2019 PingCAP, Inc.
```
Comment (WangXiangUSTC) [inline dumpling/dumpling.go]:
```suggestion
// Copyright 2020 PingCAP, Inc.
```

File: dumpling/dumpling.go [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+
+	var ret []string
+	if cfg.SkipTzUTC {
+		// TODO: support skip-tz-utc
+		ret = append(ret, "--skip-tz-utc")
+	}
+
+	extraArgs := strings.Fields(cfg.ExtraArgs)
+	if len(extraArgs) > 0 {
+		ret = append(ret, ParseArgLikeBash(extraArgs)...)
```
Comment (WangXiangUSTC) [inline dumpling/dumpling.go]: do we need to parse the `extra-args`, like some config in the integration test, the user may set `extra-args: "--statement-size=100"`,  it may cause incompatible problem
